man/man1/pupsconf.1
.TH PUPSCONF 1 "21st February 2009" "PUPSP3 build tools" "PUPSP3 build tools"

.SH NAME
PUPSCONF \- Makefile configuration facility for PUPS 
.br

.SH SYNOPSIS
.B pupsconf
[
.B -usage
|
.B -help
] | [
.B
[
.B -cdir
.I configuration directory
]!
.I Makefile configuration file name 
!!
.I Template makefile name 
!!
.I Makefile name
[
.B dserial
.I serial number
] | [
.B sdongle
.I dongle file name | dongle key
]
[2> <log file>]
.br

.SH DESCRIPTION
.I pupsconf(1) 
is a utility similar to gnu configure,
.I configure(1)
which generates makefiles from a
.I makefile template
and a
.I configuration file
which contains system dependant makefile fragments to be subsituted in the
makefile.
.I pupsconfig(1)
is both more simple and less versatile than
.I configure(1),
but is useful for generating makefiles for machines/environments which require
different CFLAGS and LDFLAG definitions. The configuration information is stored
in the directory ./config and is of the form:
.br

.DS B
# comment
<macro dummy name to be subsituted 1>
<substition string 1>

# comment
<macro dummy name to be subsituted 1>
<substition string 1>

end
.DE

If the dummy name is not to be subsituted the substitution string "no" or
"none" should be used.  The
.I end
statement at the end of the configuration file is mandatory.
.br

.SH OPTIONS

.B -cdir
.I configuration file
specifies where
.I pupsconf(1)
will find the template used to generate a
.B Makefile
from a
.B Makfile.in
file.
.br

These options are only valid if
.I pupsconf(1)
has been built with
.I secure mode extensions.
.br

.B Makefile configuration file name
is the file which contains the macros to be substituted in the
.B Makefile.in
file
.br


.B Template makefile name
is the
.B Makefile.in
from which the
.B Makefile
is generated.
.br


.B Makefile name
is the name of the
.B Makefile
to be generated.
.br

.B dserial
.I disk serial number
tells
.I pupsconf(1)
to use the hard disk serial number as the software build authentication token when generating makefile. If a
serial number is specified, that is used (enable secure remote build). If not, the serial number
of the disk the software is to be built on is used. This option and
.B sdongle
are
.I mutually exclusive.
.br

.B sdongle
.I dongle file name | dongle key
tells
.I puspconf(1)
to use either the (random) inode of the supplied
.I dongle file
or the supplied
.I dongle key
as the authentication token when generating makefile. If a
.I dongle file name
is specified, it is assumed to be on the disk being used for the build:
.I dongle key
can refer to an appropriate file on the local build disk, or a remote disk, which
means that this option can be used for remote builds.
.B sdongle
and
.I dserial
are
.I mutually exclusive.
.br


.SH EXAMPLE MACRO FILE
.br

# This is the name of the macro
.br
TARGET_CFLAGS
.br
.br

#This is its body
.br
-verbose -DHAVE_MACROS
.br
.br

# End of macro definitions
.br
end
.br

.SH FILES
Makefile.in
.br
Makefile
.br
config_file
.br

.SH EXIT STATUS

255: error
.br
1:
.B -usage
or
.B -help
.br
0: no error
.br

.SH ENVIRONMENT VARIABLES
PUPSCONF gives the name of the directory to be searched for the configuration
file.

If secure build mode extensions are enabled
.B USE_DISK_SERIAL
.I disk_serial
tells
.I pupsconf(1)
to use the disk serial number as the secure authentication token for building the software.
If
.I disk_serial
is specified, it is used as the disk serial, otherwise the build disk is probed for its serial number.
This environmental option and
.B USE_SOFT_DONGLE
are mutually exclusive. Note they disk serial number must be
.I pre-encrypted.
Thus, it is generated by running the
.I hdid(1)
command on the target host:
hdid <disk device>
.br

.B USE_SOFT_DONGLE
.I dongle file name | dongle key
tells
.I pupsconf(1)
to use the (randomised) inode of the specified
.I dongle file
or the specified
.I dongle key
as the secure authentication token for building the software. The
.I dongle file
is assumed to be on the build disk, but the
.I dongle key
can refer to an appropriate file anywhere, permitting soft dongles to be used for
remote builds.  This environmental option and
.B USE_DISK_SERIAL
are mutually exclusive. Note the dongle key
.B must be specified in hex.
This key is obtained by running
.I sdongle(1)
on the target host.
.br

.SH BUGS
At present the
.I disk serial
authentication mechanism only works for
.B ATAPI
not
.B SCSI
disks.
.br

.SH COPYRIGHT
(C) Tumbling Dice Ltd.
.br

.SH PROGRAMMER
M.A. O'Neill.
 
.SH SEE ALSO
configure(1), hdid(1), sdongle(1)
.br


pupscore.src/pupsconf.c
/*-------------------------------------------------------
    Purpose: Tool to configure P3 Makefiles for different
             architectures.

    Author:  M.A. O'Neill
             Tumbling Dice Ltd
             Gosforth
             Newcastle upon Tyne
             NE3 4RT
             United Kingdom

    Version: 3.02 
    Dated:   11th December 2024 
    E-mail:  mao@tumblingdice.co.uk
------------------------------------------------------*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <string.h>
#include <bsd/string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdint.h>


/*---------*/
/* Defines */
/*---------*/
/*---------*/
/* Version */
/*---------*/

#define PUPSCONF_VERSION    "3.02"


/*-------------*/
/* String size */
/*-------------*/

#define SSIZE               2048 


/*---------------*/
/* Boolean flags */
/*---------------*/

#ifdef TRUE
#undef TRUE
#endif /* TRUE */

#ifdef FALSE 
#undef FALSE 
#endif /* FALSE */

#define TRUE      255
#define FALSE     0


/*-----------------------------------------------*/
/* Default config dir (if non supplied via make) */
/*-----------------------------------------------*/

#ifndef DEFAULT_CONFIGDIR
#define DEFAULT_CONFIGDIR "../config"
#endif /* DEFAULT_CONFIGDIR */


/*-------------------------------------------------*/
/* Functions which are private to this application */
/*-------------------------------------------------*/

static int32_t strin    (const char *, const char *);
static int32_t error    (const char *);




/*------------------*/
/* Main entry point */
/*------------------*/

int main(int argc, char *argv[])

{   uint32_t i;

    int32_t  start                     = 1,
             ddes                      = (-1),
             m_names                   = 0;

    FILE     *pstream                  = (FILE *)NULL,
             *cstream                  = (FILE *)NULL,
             *ostream                  = (FILE *)NULL;

    char     m_d_name[SSIZE]           = "",
             next_m_s_frag[SSIZE]      = "",
             m_s_name[SSIZE]           = "",
             tmpfile[SSIZE]            = "",
             sed_c_frag[SSIZE]         = "",
             sed_command[SSIZE]        = "",
             disk_serial[SSIZE]        = "",
             hd_device[SSIZE]          = "",
             config_target[SSIZE]      = "",
             m_d_namelist[256][SSIZE]  = { "" };


    /*--------------------*/
    /* Parse command line */
    /*--------------------*/

    if(argc < 4 || strcmp(argv[1],"-help") == 0 || strcmp(argv[1],"-usage") == 0)
    {  (void)fprintf(stderr,"\npupsconf version %s, (C) Tumbling Dice 2002-2024 (gcc %s: built %s %s)\n\n",PUPSCONF_VERSION,__VERSION__,__TIME__,__DATE__);
       (void)fprintf(stderr,"PUPSCONF is free software, covered by the GNU General Public License, and you are\n");
       (void)fprintf(stderr,"welcome to change it and/or distribute copies of it under certain conditions.\n");
       (void)fprintf(stderr,"See the GPL and LGPL licences at www.gnu.org for further details\n");
       (void)fprintf(stderr,"PUPSCONF comes with ABSOLUTELY NO WARRANTY\n\n");

       (void)fprintf(stderr,"\nusage: pupsconf [-usage | -help] | [-cdir <configuration directory:../config>]\n");
       (void)fprintf(stderr,"                !<target> <buildfile.in> <buildfile>!\n");


       (void)fprintf(stderr,"\nDefault config directory is: \"%s\"\n\n",DEFAULT_CONFIGDIR); 
       (void)fflush(stderr);

       exit(255);
    }

    if(strcmp(argv[1],"-cdir") == 0)
    {  start = 3;
       (void)snprintf(config_target,SSIZE,"%s/%s",argv[2],argv[3]);
    }
    else if(getenv("P3_CONFDIR") == (char *)NULL)
       (void)snprintf(config_target,SSIZE,"%s/%s",DEFAULT_CONFIGDIR,argv[1]);
    else
       (void)snprintf(config_target,SSIZE,"%s/%s",getenv("P3_CONFDIR"),argv[1]);


    #ifdef DEBUG  
    (void)fprintf(stderr,"P3_CONFDIR: %s\n",getenv("P3_CONFDIR"));
    (void)fprintf(stderr,"TARGET: %s\n",argv[1]);
    (void)fflush(stderr);
    #endif /* DEBUG */
     
    if((cstream = fopen(config_target,"r")) == (FILE *)NULL)
    {  (void)fprintf(stderr,"pupsconf: cannot find configuration file for target %s (pathname is %s)\n",argv[start],config_target);
       (void)fflush(stderr);

       (void)kill(0,SIGTERM);
       exit(255);
    }
    else
    {  (void)fprintf(stderr,"    Configuring %s for %s system\n",argv[start+1],argv[start]);
       (void)fflush(stderr);
    }


    /*---------------------------------------------------*/
    /* Check that we can actually access the target file */
    /*---------------------------------------------------*/

    if(access(argv[start + 1],R_OK) == (-1))
    {  (void)fprintf(stderr,"pupsconf: cannot find target %s\n",argv[start + 1]);
       (void)fflush(stderr);

       exit(255);
    }


    /*-----------------------------------------------------------------*/
    /* Main macro substitution loop -- get dummy name and string to be */
    /* substituted and build an appropriate sed command                */
    /*-----------------------------------------------------------------*/

    (void)strlcpy(sed_command,"sed \"",SSIZE);
    do {    do {   (void)fgets(m_d_name,SSIZE,cstream);
               } while(m_d_name[0] == '#' || m_d_name[0] == '\n');
            m_d_name[strlen(m_d_name) - 1] = '\0';

            if(strcmp(m_d_name,"end") == 0 || strcmp(m_d_name,"END") == 0)
               goto run_sed_command;


            /*------------------------------------------------------*/
            /* Move to next macro ignoring comments and empty lines */
            /*------------------------------------------------------*/

            do {   (void)fgets(next_m_s_frag,255,cstream);
               } while(next_m_s_frag[0] == '#' || next_m_s_frag[0] == '\n');


            /*---------------------------------*/
            /* Extract macro substitution text */
            /*---------------------------------*/

            next_m_s_frag[strlen(next_m_s_frag) - 1] = '\0';
            (void)strlcpy(m_s_name,next_m_s_frag,SSIZE);
            do {   (void)fgets(next_m_s_frag,255,cstream);

                   if(next_m_s_frag[0] != '#' && next_m_s_frag[0] != '\n')
                   {  next_m_s_frag[strlen(next_m_s_frag) - 1] = '\0';
                      (void)strlcat(m_s_name," ",          SSIZE);
                      (void)strlcat(m_s_name,next_m_s_frag,SSIZE);
                   }

               } while(next_m_s_frag[0] != '#' && next_m_s_frag[0] != '\n');

            for(i=0; i<m_names; ++i)
            {  if(strin(m_d_name,m_d_namelist[i]) == TRUE)
               {  (void)fprintf(stderr,"pupsconf: macro name \"%s\" (in %s) not unique\n",
                                                                          m_d_namelist[i],
                                                                            config_target);
                  (void)fflush(stderr);

                  (void)kill(0,SIGTERM);
                  exit(255);
               }
            }

            (void)strlcpy(m_d_namelist[m_names],m_d_name,SSIZE);

            if(strcmp(m_s_name,"no") == 0 || strcmp(m_s_name,"none") == 0)
               (void)snprintf(sed_c_frag,SSIZE," s!%s!!g;",m_d_name);
            else
               (void)snprintf(sed_c_frag,SSIZE," s!%s!%s!g;",m_d_name,m_s_name);


            #ifdef DEBUG
            (void)fprintf(stderr,"CMD: %s\n",sed_command);
            (void)fprintf(stderr,"FRAG: %s [ %s ] (%s)\n\n",m_d_name,m_s_name,sed_c_frag);
            (void)fflush(stderr);
            #endif /* DEBUG */

            (void)strlcat(sed_command,sed_c_frag,SSIZE);
            ++m_names;
       } while(1);

run_sed_command:

    if(access(argv[start + 2],F_OK | R_OK | W_OK) == (-1))
       (void)creat(argv[start + 2],0644);
    if((ostream = fopen(argv[start + 2],"w")) == (FILE *)NULL)
    {  (void)fprintf(stderr,"pupsconf: cannot open \"%s\"\n",ostream);
       (void)fflush(stderr);

       exit(01);
    }

    (void)fprintf(ostream,"# Build file generated automatically by pupsconf (version 1.0)\n");
    (void)fprintf(ostream,"# do not edit\n\n");

    (void)fclose(ostream);

    (void)snprintf(tmpfile,SSIZE,"\" <%s >%s",argv[start + 1],argv[start + 2]);
    (void)strlcat(sed_command,tmpfile,SSIZE);
    (void)system(sed_command);


    (void)fprintf(stderr,"    Build file for target \"%s\" generated in %s (%d macros)\n",
                                                  argv[start + 1],argv[start + 2],m_names);
    (void)fflush(stderr);

    exit(0);
}



/*------------------------------------------------------*/
/* Look for the occurence of string s2 within string s1 */
/*------------------------------------------------------*/              
             
static int32_t strin(const char *s1, const char *s2)                                                   
             
{   uint32_t i,                                                                                   
             cmp_size,
             chk_limit;

    if(strlen(s2) > strlen(s1))                                                
       return(FALSE);                                             

    chk_limit = strlen(s1) - strlen(s2) + 1;
    cmp_size  = strlen(s2);               

    for(i=0; i<chk_limit; ++i)         
       if(strncmp(&s1[i],s2,cmp_size) == 0)
          return(TRUE);                     
       
    return(FALSE);             
}




/*---------------*/
/* Error handler */
/*---------------*/

static int32_t error(const char *error_string)

{  char hostname[SSIZE] = "";

   (void)gethostname(hostname,SSIZE);
   (void)fprintf(stderr,"    ERROR pupsconf (%d@%s)]: %s\n",getpid(),hostname,error_string);
   (void)fflush(stderr);

   exit(255);
}

